lib/plugins/aws/deploy/lib/uploadCloudFormationTemplate.js

```diff
@@ -0,0 +1,32 @@
+'use strict';
+
+const BbPromise = require('bluebird');
+
+module.exports = {
+  uploadCloudFormationTemplateToS3Bucket() {
+    this.serverless.cli.log('Uploading CloudFormation file to S3...');
+
+    const body = JSON.stringify(this.serverless.service.provider.compiledCloudFormationTemplate);
+
+    const params = {
+      Bucket: this.bucketName,
+      Key: 'compiled-cloudformation-template.json',
+      Body: body,
+    };
+
+    return this.sdk.request('S3',
+      'putObject',
+      params,
+      this.options.stage,
+      this.options.region);
+  },
+
+  uploadCloudFormationTemplate() {
```

Comment:
How about for every deployment we create a separate bucket and put the zip file and the CF template in that bucket? Then the logic for removing old versions would simply be order buckets by name (Imho bucket should have a mix of timestamp and readable date, e.g. 141264711231-2016-08-18-15:48:32) so that it can be easily ordered by the first part of the timestamp but its still human readable.

And the Zip file and CF template in each bucket could have the same name and if we want to add more artefacts in the future thats simple (e.g. nested stacks)